internal/server/auditsink/auditsink.go

```diff
@@ -0,0 +1,97 @@
+package auditsink
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+// AuditEvent holds information that represents an audit internally.
+type AuditEvent struct {
+	ResourceName string `json:"resourceName"`
+}
+
+// AuditSink is the abstraction for various audit sink configurations
+// that Flipt will support.
+type AuditSink interface {
+	SendAudits([]AuditEvent) error
+	fmt.Stringer
+}
+
+// Publisher holds information about the configured sinks that we are going
+// to send audit events to.
+type Publisher struct {
+	mtx         sync.Mutex
+	logger      *zap.Logger
+	bufferSize  int
+	sinks       []AuditSink
+	auditEvents []AuditEvent
+	ticker      *time.Ticker
+}
+
+// NewPublisher is the constructor for a Publisher.
+func NewPublisher(logger *zap.Logger, bufferSize int, sinks []AuditSink, tickerDuration time.Duration) *Publisher {
+	p := &Publisher{
+		logger:      logger,
+		bufferSize:  bufferSize,
+		sinks:       sinks,
+		auditEvents: make([]AuditEvent, 0),
+		ticker:      time.NewTicker(tickerDuration),
+	}
+
+	go p.flushWhenNecessary()
+
+	return p
+}
+
+// flush flushes the buffer to the configured sinks.
+func (p *Publisher) flush(auditEvent *AuditEvent) {
+	copiedEvents := make([]AuditEvent, len(p.auditEvents))
+	copy(copiedEvents, p.auditEvents)
+	p.auditEvents = make([]AuditEvent, 0)
+	if auditEvent != nil {
+		p.auditEvents = append(p.auditEvents, *auditEvent)
+	}
+	p.mtx.Unlock()
+	for _, sink := range p.sinks {
+		go func(s AuditSink) {
+			err := s.SendAudits(copiedEvents)
+			if err != nil {
+				p.logger.Warn("Failed to send audits to sink", zap.Stringer("sink", s))
+			}
+		}(sink)
+	}
+}
+
+// flushWhenNecessary flushes the buffer to the configured sinks if a tick elapses
+// and there are elements in the buffer, to prevent things from staying in the buffer
+// for an indefinite amount of time.
+func (p *Publisher) flushWhenNecessary() {
+	for {
+		<-p.ticker.C
+		p.mtx.Lock()
```

Comment:
@GeorgeMac Haha it's funny because all the bits you commented on were things I did not love about the first go round of commits. I like this idea, it's clearer, and all the locking and unlocking are happening in one place.

Reply:
Nice 😄 you probably would’ve got to it if I had been patient and waited for you to undraft 😆